d2/src/main/java/com/linkedin/d2/balancer/servers/ZooKeeperAnnouncer.java

```diff
@@ -211,8 +248,88 @@ public void onSuccess(None result)
         }
         runNextMarkUpOrMarkDown();
       }
-    });
+    };
+
+    final Callback<None> warmUpMarkDownCallback = new Callback<None>()
+    {
+      @Override
+      public void onError(Throwable e)
+      {
+        // It is important here to retry the markDown for warm-up cluster.
+        // We cannot go ahead to markUp the regular cluster, as the warm-up cluster to uris association has not been deleted
+        // from the zookeeper store.
+        if (e instanceof KeeperException.ConnectionLossException || e instanceof KeeperException.SessionExpiredException)
+        {
+          _log.warn("failed to markDown uri {} on warm-up cluster {} due to {}.", _uri, _warmupClusterName, e.getClass().getSimpleName());
+          // Setting to null because if that connection dies, when don't want to continue making operations before
+          // the connection is up again.
+          // When the connection will be up again, the ZKAnnouncer will be restarted and it will read the _isWarmingUp
+          // value and mark down warm-up cluster again if necessary
+          _nextOperation = null;
+          _isRunningMarkUpOrMarkDown = false;
+        }
+        else
+        {
+          callback.onError(e);
+          //continue to mark up to the regular cluster
+          _server.markUp(_cluster, _uri, _partitionDataMap, _uriSpecificProperties, markUpCallback);
+        }
+      }
+
+      @Override
+      public void onSuccess(None result)
+      {
+        // Mark _isWarmingUp to true to indicate warm up is completed
+        _isWarmingUp = false;
+        // Note that the pending callbacks we see at this point are from the requests that are filed before us because
+        // zookeeper guarantees the ordering of callback being invoked.
+        synchronized (ZooKeeperAnnouncer.this)
+        {
+          // drain _pendingMarkDown with successful result.
```

Comment:
Should this comment reference _pendingWarmUpMarkDown, not _pendingMarkDown?  The code below it is draining _pendingWarmUpMarkDown.

Reply:
Thank you for spotting this! Fixed.